### PR TITLE
Expose `parse_backend_addr` for `GUI`

### LINKED ISSUE #5173: Expose method to create `BackendAddrBootstrapOrganization` from GUI
`bootstrap_organization` takes a `BackendOrganizationBootstrapAddr` as parameter, but currently the only way to generate this type (which is in fact a string) is to concatenate by hand.

Instead we should have a factory function:
```javascript
var domain = "example.com";
var port = 8080; // can also be null
var organization_id = "CoolOrg";
var addr = await libparsec.buildBackendOrganizationBootstrapAddr(domain, port, organization_id);

// Then bootstrap
var available_device = await libparsec.bootstsrapOrganization([...], addr, [...]);
```

To keep things simple, the GUI is responsible to provide `domain` and `port`: `buildBackendOrganizationBootstrapAddr` returns specific errors if they are invalid e.g. `port=0` or `domain=""`, this way the GUI can use this function to also validate the user input with `buildBackendOrganizationBootstrapAddr` (it only has to split the user input at the first `:` character, then call `buildBackendOrganizationBootstrapAddr` and display feedback depending on the type of error returned)

### PR DESCRIPTION
**Pull request by TimeEngineer**
## Describe your changes

closes #5173 

## Checklist

Before you submit this pull request, please make sure to:

- [x] Keep changes in the pull request as small as possible
   
- [x] Ensure the commit history is sanitized
   
- [x] Give a meaningful title to your PR
- [x] Describe your changes
   
- [x] Link any related issue in the description
   
- [x] Link any dependent pull request in the description

If this PR adds a new feature or fixes a bug:

- [ ] I have added or updated relevant tests

If this PR modifies installation or updates packages:

- [ ] I have checked that installation works on all platforms (Windows, Linux, Mac ...)
    

- [ ] I have updated related documentations (installation commands, fix to specific problems ...)

If this PR is related to GUI:

- [ ] I have checked that GUIs affected by my changes are working properly.

If this PR affects the user:

- [ ] I have created a news fragment
   
- [ ] I have updated the user documentation

### CODE REVIEW

File: libparsec/src/lib.rs [hunk 1]
```diff
@@ -31,3 +31,14 @@ pub use testbed::*;
 pub async fn list_available_devices(config_dir: &std::path::Path) -> Vec<AvailableDevice> {
     libparsec_platform_device_loader::list_available_devices(config_dir).await
 }
+
+pub async fn build_backend_addr(hostname: String, port: Option<u32>, use_ssl: bool) -> BackendAddr {
```
Comment (FirelightFlagboy) [inline libparsec/src/lib.rs:35]:
```suggestion
pub fn build_backend_addr(hostname: String, port: Option<u32>, use_ssl: bool) -> BackendAddr {
```

Don't need to be async

File: libparsec/src/lib.rs [hunk 2]
```diff
@@ -31,3 +31,14 @@ pub use testbed::*;
...
+    BackendAddr::new(hostname, port.map(|x| x as u16), use_ssl)
+}
+
+pub async fn build_backend_organization_bootstrap_addr(
```
Comment (FirelightFlagboy) [inline libparsec/src/lib.rs:39]:
```suggestion
pub fn build_backend_organization_bootstrap_addr(
```

Don't need to be async

File: bindings/generator/api/common.py [hunk 3]
```diff
@@ -199,3 +199,15 @@ class UserProfile(Variant):
     Admin = VariantItemUnit
     Standard = VariantItemUnit
     Outsider = VariantItemUnit
+
+
+async def build_backend_addr(
```
Comment (FirelightFlagboy) [inline bindings/generator/api/common.py:204]:
```suggestion
def build_backend_addr(
```

Don't need to be async

File: bindings/generator/api/common.py [hunk 4]
```diff
@@ -199,3 +199,15 @@ class UserProfile(Variant):
...
+    hostname: str, port: Optional[U32BasedType], use_ssl: bool
+) -> BackendAddr:
+    raise NotImplementedError
+
+
+async def build_backend_organization_bootstrap_addr(
```
Comment (FirelightFlagboy) [inline bindings/generator/api/common.py:210]:
```suggestion
def build_backend_organization_bootstrap_addr(
```

Don't need to be async

**Review by FirelightFlagboy — COMMENTED** [review]
Maybe we should consider to create a separate `generator/api/addr.py` :thinking: (maybe for a latter PR) WDYT ?

**Comment by TimeEngineer** [comment]
> Maybe we should consider to create a separate `generator/api/addr.py` 🤔 (maybe for a latter PR) WDYT ?

It's a good idea

**Review by FirelightFlagboy — APPROVED** [review]

File: client/src/plugins/libparsec/definitions.d.ts [hunk 5]
```diff
@@ -752,6 +775,9 @@ export interface LibParsecPlugin {
         recipient: UserID,
         role: RealmRole | null
     ): Promise<Result<null, ClientWorkspaceShareError>>
+    getBackendAddrType(
+        url: string
+    ): Promise<Result<BackendAddrType, string>>
```
Comment (Max-7) [inline client/src/plugins/libparsec/definitions.d.ts:780]: What's the `string` here? Shouldn't it be an error type?

File: client/src/plugins/libparsec/definitions.d.ts [hunk 6]
```diff
@@ -320,6 +320,29 @@ export type UserProfile =
   | UserProfileOutsider
   | UserProfileStandard
 
+// BackendAddrType
+export interface BackendAddrTypeInvitation {
+    tag: 'Invitation'
+}
+export interface BackendAddrTypeNormal {
```
Comment (Max-7) [inline client/src/plugins/libparsec/definitions.d.ts:327]: I don't like the word `Normal` here, it sounds weird. The other ones are not abnormal, they're just specializations. What about `BackendAddrTypeBase` instead?

**Review by Max-7 — CHANGES_REQUESTED** [review]
Missing a function to generate a `BackendAddrTypeOrganizationBootstrap` from a `BackendAddrBase` and an `OrganizationID`.

File: client/src/plugins/libparsec/definitions.d.ts (hunk 5 above)
Comment (TimeEngineer) [reply client/src/plugins/libparsec/definitions.d.ts:784]: It's any url which will be parsed to know which BackendAddr is

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: bindings/generator/api/common.py [hunk 7]
```diff
@@ -199,3 +199,15 @@ class UserProfile(Variant):
     Admin = VariantItemUnit
     Standard = VariantItemUnit
     Outsider = VariantItemUnit
+
+
+class BackendAddrType(Variant):
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply bindings/generator/api/common.py]: Done

File: libparsec/src/lib.rs [hunk 8]
```diff
@@ -31,3 +31,112 @@ pub use testbed::*;
 pub async fn list_available_devices(config_dir: &std::path::Path) -> Vec<AvailableDevice> {
     libparsec_platform_device_loader::list_available_devices(config_dir).await
 }
+
+pub enum ParseBackendAddrError {
```
Comment (touilleMan) [inline libparsec/src/lib.rs:35]: put this in `libparsec/src/addr.rs` ?

File: libparsec/crates/types/src/addr.rs [hunk 9]
```diff
@@ -26,6 +26,89 @@ const PARSEC_NO_SSL_DEFAULT_PORT: u16 = 80;
 /// see. https://github.com/servo/rust-url/issues/763
 struct ParsecUrlAsHTTPScheme(pub Url);
 
+impl ParsecUrlAsHTTPScheme {
```
Comment (touilleMan) [inline libparsec/crates/types/src/addr.rs:29]: You were right, that's a pretty good refacto :+1:

**Review by touilleMan — APPROVED** [review]

File: libparsec/src/lib.rs (hunk 8 above)
Comment (TimeEngineer) [reply libparsec/src/lib.rs]: Done

**Review by Max-7 — APPROVED** [review]

File: bindings/generator/api/addr.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    custom_to_rs_string = "|addr: libparsec::BackendOrganizationBootstrapAddr| -> Result<String, &'static str> { Ok(addr.to_url().into()) }"
+
+
+class BackendInvitationAddr(StrBasedType):
+    custom_from_rs_string = "|s: String| -> Result<_, String> { libparsec::BackendInvitationAddr::from_any(&s).map_err(|e| e.to_string()) }"
+    custom_to_rs_string = "|addr: libparsec::BackendInvitationAddr| -> Result<String, &'static str> { Ok(addr.to_url().into()) }"
+
+
+class ParseBackendAddrError(Variant):
+    InvalidUrl = VariantItemUnit
```
Comment (FirelightFlagboy) [inline bindings/generator/api/addr.py:39]: Use an enum ?

Comment (TimeEngineer) [reply bindings/generator/api/addr.py]: Done

File: bindings/generator/api/addr.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    custom_to_rs_string = "|addr: libparsec::BackendOrganizationBootstrapAddr| -> Result<String, &'static str> { Ok(addr.to_url().into()) }"
+
+
+class BackendInvitationAddr(StrBasedType):
+    custom_from_rs_string = "|s: String| -> Result<_, String> { libparsec::BackendInvitationAddr::from_any(&s).map_err(|e| e.to_string()) }"
+    custom_to_rs_string = "|addr: libparsec::BackendInvitationAddr| -> Result<String, &'static str> { Ok(addr.to_url().into()) }"
+
+
+class ParseBackendAddrError(Variant):
+    InvalidUrl = EnumItemUnit
```
Comment (FirelightFlagboy) [inline bindings/generator/api/addr.py:39]: As an afterthought, `ParseBackendAddrError` should be `ErrorVariant` like other errors :thinking:

Comment (TimeEngineer) [reply bindings/generator/api/addr.py:39]: True

File: bindings/generator/api/addr.py (hunk 10 above)
Comment (touilleMan) [reply bindings/generator/api/addr.py]: good catch ! this should be an `ErrorVariant`, not a `Variant`

Comment (FirelightFlagboy) [reply bindings/generator/api/addr.py]: Did you mean https://github.com/Scille/parsec-cloud/pull/5182#discussion_r1324359587 ?

**Pull request merged**